File: mergetiles.pl
#!/usr/bin/perl

# Merge two tile sets, blending tiles together when both exist
# Copyright (C) 2013  Jesse McGraw (jlmcgraw@gmail.com)
#
#-----------------------------------------------------------------------------
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see [http://www.gnu.org/licenses/].

#Standard modules
use strict;
use warnings;
use autodie;
use Carp;

# Using this so users don't need to globally install modules
# allows using "carton install" instead
use FindBin '$Bin';
use lib "$FindBin::Bin/local/lib/perl5";

# Non-Standard modules that should be installed locally
use Modern::Perl;
use Params::Validate qw(:all);
use File::Slurp;
use File::Copy;

#Call the main routine and exit with its return code
exit main(@ARGV);

sub main {
    
    # Get the source tiles directory argument
    my $source_directory = $ARGV[0];
    
    # Get the merged tiles directory argument 
    my $base_directory    = $ARGV[1];

    print STDOUT "Merging tiles from $source_directory into $base_directory\n";

    my @areas = read_dir($source_directory);
   
    foreach my $area (@areas) {
        my $overlay_directory = "$source_directory/$area"; 

        print STDOUT "processing tiles: $area\n";
        
        # Get all of the zoom level subdirectories in the $overlay_directory
        my @overlay_zoom_levels = read_dir($overlay_directory);

        foreach my $zoomlevel (@overlay_zoom_levels) {
            
            if ( -d "$overlay_directory/$zoomlevel" ) {

                # Make the base/destination directory if it doesn't exist
                unless ( -e "$base_directory/$zoomlevel" ) {
                    mkdir "$base_directory/$zoomlevel";
                }

                # For each x level...
                my @overlay_x_levels = read_dir("$overlay_directory/$zoomlevel");

                foreach my $x (@overlay_x_levels) {
                    # Make the base/destination directory if it doesn't exist
                    if ( -d "$overlay_directory/$zoomlevel/$x" ) {
                        unless ( -e "$base_directory/$zoomlevel/$x" ) {
                            mkdir "$base_directory/$zoomlevel/$x";
                        }

                        # For each tile...
                        my @overlay_y_tiles = read_dir("$overlay_directory/$zoomlevel/$x");

                        foreach my $y (@overlay_y_tiles) {
                            # If the both base and overlay tiles exist, then call the ImageMagick "convert" function to composite the images. 
                            if ( -e "$base_directory/$zoomlevel/$x/$y" ) {
                                qx(convert "$base_directory/$zoomlevel/$x/$y" "$overlay_directory/$zoomlevel/$x/$y" -composite "$base_directory/$zoomlevel/$x/$y");
                            }
                            # Otherwise do a regular copy from overlay tile to base directory
                            else {
                                move("$overlay_directory/$zoomlevel/$x/$y", "$base_directory/$zoomlevel/$x/$y");
                            }
                        }
                    }
                }
            }     
        }
    }
}

File: mergetiles.py
#!/usr/bin/env python3

import os
import sys
import shutil
from multiprocessing import Pool

def main(): 
    # Get the passed in maximum number of processes argument
    max_processes = int(sys.argv[1])

    # Get the source tiles directory argument
    source_tiles_directory = sys.argv[2]

    # Get the merged tiles directory argument 
    base_tiles_directory = sys.argv[3]

    # Read the source tiles directory
    areas = os.listdir(source_tiles_directory)

    with Pool(processes=max_processes) as pool:
        pool.starmap(process_zoom_levels, [(base_tiles_directory, os.path.join(source_tiles_directory, area)) for area in areas])

    #for area in areas:
    #    process_zoom_levels (base_tiles_directory, os.path.join(source_tiles_directory, area))

def process_zoom_levels(base_directory, overlay_directory):
    # Get all of the zoom level subdirectories in the overlay_directory
    overlay_tiles_zoom_levels = os.listdir(overlay_directory)

    for zoomlevel in overlay_tiles_zoom_levels:
        if os.path.isdir(os.path.join(overlay_directory, zoomlevel)):
            # Make the base/destination directory if it doesn't exist
            base_zoomlevel_path = os.path.join(base_directory, zoomlevel)
            if not os.path.exists(base_zoomlevel_path):
                os.makedirs(base_zoomlevel_path)

            # For each column...
            overlay_tiles_x_levels = os.listdir(os.path.join(overlay_directory, zoomlevel))

            for x in overlay_tiles_x_levels:
                # Make the base/destination directory if it doesn't exist
                overlay_x_path = os.path.join(overlay_directory, zoomlevel, x)
                if os.path.isdir(overlay_x_path):
                    base_x_path = os.path.join(base_zoomlevel_path, x)
                    if not os.path.exists(base_x_path):
                        os.makedirs(base_x_path)

                    # For each tile...
                    overlay_tiles_y_tiles = os.listdir(overlay_x_path)

                    for y in overlay_tiles_y_tiles:
                        base_y_path = os.path.join(base_x_path, y)
                        overlay_y_path = os.path.join(overlay_x_path, y)
                        # If both base and overlay tiles exist then composite them together with "convert"
                        if os.path.exists(base_y_path):
                            os.system(f'convert "{base_y_path}" "{overlay_y_path}" -composite "{base_y_path}"')
                        # Otherwise do a regular copy from overlay tile to base directory
                        else:
                            shutil.copy(overlay_y_path, base_y_path)

if __name__ == "__main__":   
    main()
